packages/server/lib/browsers/bidi_automation.ts

```diff
@@ -0,0 +1,277 @@
+    // so the request-middleware can identify the request
+
+    if (params.isBlocked) {
+      params.request.headers.push({
+        name: 'X-Cypress-Is-WebDriver-BiDi',
+        value: {
+          type: 'string',
+          value: 'true',
+        },
+      })
+
+      if (params.context === this.autContextId && resourceType === 'document') {
+        debug(`AUT request detected, adding X-Cypress-Is-AUT-Frame for request ID: ${params.request.request}`)
+
+        params.request.headers.push({
+          name: 'X-Cypress-Is-AUT-Frame',
+          value: {
+            type: 'string',
+            value: 'true',
+          },
+        })
+      }
+
+      try {
+        debug(`continuing request ID: ${params.request.request}`)
+
+        await this.webDriverClient.networkContinueRequest({
+          request: params.request.request,
+          headers: params.request.headers,
+          cookies: params.request.cookies,
+        })
+      } catch (err: unknown) {
+        // happens if you kill the Cypress app in the middle of request interception. This error can be ignored
+        if (!(err as Error)?.message.includes('no such request')) {
+          throw err
+        }
+      }
+    }
+  }
+
+  private onResponseStarted = (params: NetworkResponseStartedParameters) => {
+    debugVerbose('received network.responseStarted %o', params)
+
+    if (params.response.fromCache) {
+      this.automation.onRemoveBrowserPreRequest?.(params.request.request)
+    }
+  }
+
+  private onResponseComplete = (params: NetworkResponseCompletedParameters) => {
+    debugVerbose('received network.responseComplete %o', params)
+
+    if (params.response.fromCache) {
+      this.automation.onRemoveBrowserPreRequest?.(params.request.request)
+
+      return
+    }
+
+    const parsedHeaders = {}
+
+    params.response.headers.forEach((header) => {
+      parsedHeaders[header.name] = header.value.value
+    })
+
+    const browserResponseReceived: BrowserResponseReceived = {
+      requestId: params.request.request,
+      status: params.response.status,
+      headers: parsedHeaders,
+    }
+
+    this.automation.onRequestEvent?.('response:received', browserResponseReceived)
+  }
+
+  private onFetchError = (params: NetworkFetchErrorParameters) => {
+    debugVerbose('received network.fetchError %o', params)
+
+    this.automation.onRemoveBrowserPreRequest?.(params.request.request)
+  }
+
+  close () {
+    this.webDriverClient.off('network.beforeRequestSent', this.onBeforeRequestSent)
+    this.webDriverClient.off('network.responseStarted', this.onResponseStarted)
+    this.webDriverClient.off('network.responseCompleted', this.onResponseComplete)
+    this.webDriverClient.off('network.fetchError', this.onFetchError)
+    this.webDriverClient.off('browsingContext.contextCreated', this.onBrowsingContextCreated)
+    this.webDriverClient.off('browsingContext.contextDestroyed', this.onBrowsingContextDestroyed)
```

Comment:
Do we need to unsubscribe to `BIDI_EVENTS` here as well?

Reply:
We could but its a bit hard to accomplish. We could probably do it [instance kill](https://github.com/cypress-io/cypress/blob/feat/implement_bidi/packages/server/lib/browsers/firefox.ts#L709) but the events effectively get unsubscribed when geckodriver and firefox are killed AFAIK